src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -1328,14 +1345,161 @@ private ItemStack forceItemStackToProvider(@NotNull final ICapabilityProvider pr
+     *
+     * @return {@link CitizenData} of the current building
+     */
+    public List<CitizenData> getAssignedCitizen()
+    {
+        return new ArrayList<>(assignedCitizen);
     }
 
+    /**
+     * Method to remove a citizen.
+     *
+     * @param citizen Citizen to be removed.
+     */
+    public void removeCitizen(final CitizenData citizen)
+    {
+        if (isCitizenAssigned(citizen))
+        {
+            assignedCitizen.remove(citizen);
+            citizen.setWorkBuilding(null);
+            markDirty();
+        }
+    }
+
+    /**
+     * Returns if the {@link CitizenData} is the same as the worker.
+     *
+     * @param citizen {@link CitizenData} you want to compare
+     * @return true if same citizen, otherwise false
+     */
+    public boolean isCitizenAssigned(final CitizenData citizen)
+    {
+        return assignedCitizen.contains(citizen);
+    }
+
+    /**
+     * Returns the first worker in the list.
+     *
+     * @return the EntityCitizen of that worker.
+     */
+    public Optional<EntityCitizen> getMainCitizenEntity()
+    {
+        if (assignedCitizen.isEmpty())
+        {
+            return Optional.empty();
+        }
+        return assignedCitizen.get(0).getCitizenEntity();
+    }
+
+    /**
+     * Returns whether or not the building has a worker.
+     *
+     * @return true if building has worker, otherwise false.
+     */
+    public boolean hasAssignedCitizen()
+    {
+        return !assignedCitizen.isEmpty();
+    }
+
+    /**
+     * Returns the {@link net.minecraft.entity.Entity} of the worker.
+     *
+     * @return {@link net.minecraft.entity.Entity} of the worker
+     */
+    @Nullable
+    public List<Optional<EntityCitizen>> getAssignedEntities()
+    {
+        return assignedCitizen.stream().filter(Objects::nonNull).map(CitizenData::getCitizenEntity).collect(Collectors.toList());
+    }
+
+    /**
+     * Assign the citizen to the current building.
+     *
+     * @param citizen {@link CitizenData} of the worker
+     */
+    public boolean assignCitizen(final CitizenData citizen)
+    {
+        if (assignedCitizen.contains(citizen))
+        {
+            return false;
+        }
+
+        // If we set a worker, inform it of such
+        if (citizen != null)
+        {
+            assignedCitizen.add(citizen);
+        }
+
+        markDirty();
+        return true;
+    }
+
+    /**
+     * Assign the citizen after loading it from NBT.
+     * @param data the citizen data.
+     */
+    public void assignCitizenFromNBtAction(final CitizenData data)
+    {
+        /**
+         * Specific classes will override this.
+         */
+    }
+
+    /**
+     * Returns whether the citizen has this as home or not.
+     *
+     * @param citizen Citizen to check.
+     * @return True if citizen lives here, otherwise false.
+     */
+    public boolean hasAssignedCitizen(final CitizenData citizen)
+    {
+        return assignedCitizen.contains(citizen);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') This call to "contains()" may be a performance hot spot if the collection is large. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)